ext/opcache/jit/zend_jit_ir.c

```diff
@@ -10047,8 +10142,12 @@ static int zend_jit_do_fcall(zend_jit_ctx *jit, const zend_op *opline, const zen
 		// JIT: EG(current_execute_data) = execute_data;
 		ir_STORE(jit_EG(current_execute_data), rx);
 
-		if (ZEND_OBSERVER_ENABLED) {
-			ir_CALL_1(IR_VOID, ir_CONST_FC_FUNC(zend_observer_fcall_begin), rx);
+		bool may_have_observer = ZEND_OBSERVER_ENABLED && (!func || (func->common.fn_flags & (ZEND_ACC_CALL_VIA_TRAMPOLINE | ZEND_ACC_GENERATOR)) == 0);
+		if (may_have_observer) {
+			ir_ref observer_handler;
+			struct jit_observer_fcall_is_unobserved_data unobserved_data = jit_observer_fcall_is_unobserved_start(jit, func, &observer_handler, rx, func_ref ? func_ref : ir_LOAD_A(jit_CALL(rx, func)));
+			jit_observer_fcall_begin(jit, rx, observer_handler);
+			jit_observer_fcall_is_unobserved_end(jit, &unobserved_data);
```

Comment:
I don't understand what do you do in ``zend_jit_push_call_frame()`` and ``zend_jit_do_fcall()`` now.
These changes are not needed for the frameless/observer support anymore.
Please remove them.

Reply:
I think you might have forgotten that there are two parts to this PR: optimizing observers in general and enabling frameless calls with observers active. See e.g. https://github.com/php/php-src/pull/13649#issuecomment-2018880983 where I discuss the impact of only generic optimizations (like this one here) and frameless support included.

Reply:
I.e. for normal, non-frameless fcalls, it's also much cheaper to JIT the runtime cache lookup than doing it in non-specialized functions.